Add support for Stochastically Perturbed Parameterizations (SPP) in FV3 and add the FV3_RRFS_v1alpha SDF.

JeffBeck-NOAA opened the pull request:
## Description
This PR adds the necessary code in fv3atm to allow for Stochastically Perturbed Parameterizations (SPP) in a set of RAP/HRRR-based physics parameterizations.  Specific to the fv3atm repository, code in this PR defines the necessary variables associated with various SPP-related fields (e.g., logical to activate SPP, parameterization-specific SPP variables, etc.) that are then passed to ccpp-physics.

The FV3_RRFS_v1alpha SDF was added again, and will not impact model/regression testing.

### Issue(s) addressed
This PR address [Issue #978](https://github.com/ufs-community/ufs-weather-model/issues/978) in ufs-weather-model.

## Testing
Regression testing for ufs-weather-model was conducted on Hera and Cheyenne.  All passed, but not bit-for-bit for RAP and HRRR cases (tests 41 and 42 on cheyenne, 44 and 45 on hera); although unlikely that this B4B discrepancy is related to code in this PR. Debugging revealed several errors in the GSL GWD interfaces, might be the cause of the B4B issue (unallocated, mis-matched array sizes, etc.) that are caught with a DEBUG compile.  These cases do not run in debug mode, and there are no debug tests in the RT, so maybe this is a known issue?  The unique element for these two cases is the GSL GWD scheme (drag_suite.F90).  @DomHeinzeller described an optimization/bug in RRTMGP that caused B4B issues, maybe reducing optimization on drag_suite.F90 would fix the issue?

## Dependencies
Depends on the following PRs from [ufs-weather-model](https://github.com/ufs-community/ufs-weather-model/pull/982), [ccpp-physics](https://github.com/NCAR/ccpp-physics/pull/820), and [stochastic_physics](https://github.com/noaa-psd/stochastic_physics/pull/51).

@llpcarson, @jwolff-ncar, @willmayfield, @bluefinweiwei, @michelleharrold, @judithberner, @pjpegion

Commit:
submodule pointer for ccpp-physics

Commit:
update submodules

Commit:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

Commit:
update submodule atmos_cubed_sphere

Commit:
Add SPP option to several physics parameterizations

Commit:
bug fixes

Commit:
Merge branch 'develop' of https://github.com/noaa-emc/fv3atm into feature/stoch_spp

Commit:
update rte/rrtmgp

Commit:
update submodule pointer

Commit:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

Commit:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

JeffBeck-NOAA committed:
Update pointer for ccpp-physics to JeffBeck-NOAA fork

climbfuji commented on atmos_model.F90:
```diff
@@ -262,7 +262,7 @@ subroutine update_atmos_radiation_physics (Atmos)
 
       if (GFS_Control%do_sppt .or. GFS_Control%do_shum .or. GFS_Control%do_skeb .or. &
-          GFS_Control%lndp_type > 0  .or. GFS_Control%do_ca ) then
+          GFS_Control%lndp_type > 0  .or. GFS_Control%do_ca .or. GFS_Control%do_spp) then
```
Probably a stupid question, but ... what is the difference between spp and sppt?

JeffBeck-NOAA replied:
Not a stupid question!  SPP is Stochastically Perturbed Parameterizations, which perturbs specific unknown parameters (or those with a known range of uncertainty) within the physics call; in this instance, for a number of RAP/HRRR-based parameterizations.  SPPT is Stochastically Perturbed Parameterization Tendencies, which perturbs all summed tendencies right after the physics call.  The latter is known as an 'ad-hoc' scheme, since it's more-or-less a white noise technique added after the fact, while the former is much more targeted, and "active", since the perturbations occur within the physics parameterizations themselves, but generally with a much smaller impact unless many parameters are perturbed across all parameterizations.

climbfuji replied:
Thanks for educating me. What a difference one letter can make.

climbfuji commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -2236,6 +2236,46 @@
   kind = kind_phys
   active = (flag_for_stochastic_skeb_option)
+[spp_wts_pbl]
+  standard_name = weights_for_stochastic_spp_pbl_perturbation
```
@grantfirl Can you please check the standard names for the newly defined variables? Do they match the standard name rules defined in https://github.com/ESCOMP/CCPPStandardNames/blob/main/StandardNamesRules.rst ?

climbfuji commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -2186,6 +2186,61 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
       enddo
     endif
 
+    idx = idx + 1
```
This entire section 2189-2243 must be in an `if` section that is only executed if those variables are allocated

JeffBeck-NOAA replied:
Fixed!  Thanks for catching this.

climbfuji commented on fv3_cap.F90:
```diff
@@ -803,6 +803,7 @@ subroutine InitializeAdvertise(gcomp, rc)
       if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
     end do
 
+    if(mype==0) print *,'in fv3_cap, aft import, bf export fields in atmos'
```
Please remvove

JeffBeck-NOAA replied:
Removed.

climbfuji commented on stochastic_physics/stochastic_physics_wrapper.F90:
```diff
@@ -91,23 +92,26 @@ subroutine stochastic_physics_wrapper (GFS_Control, GFS_Data, Atm_block, ierr)
     levs   = GFS_Control%levs
     maxblk = maxval(GFS_Control%blksz)
     nblks  = Atm_block%nblks
+    print *,"SPP: levs,nblks",levs,nblks
```
Please remove

JeffBeck-NOAA replied:
Removed.

climbfuji commented on stochastic_physics/stochastic_physics_wrapper.F90:
```diff
@@ -91,23 +92,26 @@ subroutine stochastic_physics_wrapper (GFS_Control, GFS_Data, Atm_block, ierr)
             xlat(nb,1:GFS_Control%blksz(nb)) = GFS_Data(nb)%Grid%xlat(:)
             xlon(nb,1:GFS_Control%blksz(nb)) = GFS_Data(nb)%Grid%xlon(:)
          end do
+    print *,"SPP: call init_stoch",GFS_Control%do_spp
```
No clutter please

JeffBeck-NOAA replied:
Removed.

climbfuji commented on stochastic_physics/stochastic_physics_wrapper.F90:
```diff
@@ -188,10 +196,11 @@ subroutine stochastic_physics_wrapper (GFS_Control, GFS_Data, Atm_block, ierr)
     else initalize_stochastic_physics
-      if (GFS_Control%do_sppt .OR. GFS_Control%do_shum .OR. GFS_Control%do_skeb .OR. (GFS_Control%lndp_type == 2) ) then
+      if (GFS_Control%do_sppt .OR. GFS_Control%do_shum .OR. GFS_Control%do_skeb .OR. (GFS_Control%lndp_type == 2) .OR. GFS_Control%do_spp) then
+         print *,"SPP: before run_stoch call"
```
Remove

JeffBeck-NOAA replied:
Removed.

climbfuji commented on stochastic_physics/stochastic_physics_wrapper.F90:
```diff
@@ -209,6 +218,38 @@ subroutine stochastic_physics_wrapper (GFS_Control, GFS_Data, Atm_block, ierr)
+            DO v=1,GFS_Control%n_var_spp
+               select case (trim(GFS_Control%spp_var_list(v)))
+               case('pbl')
+       print *,"SPP: spp_wts_pbl is ", v, maxval(spp_wts(:,:,:,v))
```
Remove

JeffBeck-NOAA replied:
Removed.

climbfuji commented on stochastic_physics/stochastic_physics_wrapper.F90:
```diff
@@ -78,7 +79,7 @@ subroutine stochastic_physics_wrapper (GFS_Control, GFS_Data, Atm_block, ierr)
     integer,                  intent(out)   :: ierr
 
-    integer :: nthreads, nb, levs, maxblk, nblks
+    integer :: nthreads, nb, levs, maxblk, nblks, v
```
Integer variables are usually i,j,k,l,m,n - can you use a more common name than `v` please. This convention comes from the old days where implicit variables where integers when between i and n, but reals otherwise (https://docs.oracle.com/cd/E19957-01/805-4939/z40007365fbc/index.html).

JeffBeck-NOAA replied:
Would "var" be allowed here instead of "v"?

climbfuji replied:
Yes, that provides more meaning than just one letter. If one letter representing an integer, I would use `n`.

JeffBeck-NOAA replied:
OK, given it's an integer, I changed it to "n."

climbfuji commented on stochastic_physics/stochastic_physics_wrapper.F90:
```diff
@@ -209,6 +218,38 @@ subroutine stochastic_physics_wrapper (GFS_Control, GFS_Data, Atm_block, ierr)
+       print *,"SPP: spp_wts_pbl is ", v, maxval(spp_wts(:,:,:,v))
+                 do nb=1,Atm_block%nblks
+                     GFS_Data(nb)%Coupling%spp_wts_pbl(:,:) = spp_wts(nb,1:GFS_Control%blksz(nb),:,v)
+                     GFS_Data(nb)%Intdiag%spp_wts_pbl(:,:) = spp_wts(nb,1:GFS_Control%blksz(nb),:,v)
```
See my comment above about the unnecessary Intdiag variables.

JeffBeck-NOAA replied:
Removed references to Intdiag variables.

climbfuji commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -2186,6 +2186,61 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
+    ExtDiag(idx)%mod_name = 'gfs_phys'
+    allocate (ExtDiag(idx)%data(nblks))
+    do nb = 1,nblks
+      ExtDiag(idx)%data(nb)%var3 => IntDiag(nb)%spp_wts_pbl(:,:)
```
We should be able to just point to the `Coupling` variables here.

JeffBeck-NOAA replied:
Removed the Intdiag variable references.

climbfuji commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -6506,6 +6549,16 @@ subroutine diag_create (Diag, IM, Model)
     allocate (Diag%tdomzr   (IM))
     allocate (Diag%tdomip   (IM))
     allocate (Diag%tdoms    (IM))
+    allocate (Diag%skebu_wts(IM,Model%levs))
```
Why do we need all these variable duplicates in the Intdiag type? I remember that recently @pjpegion cleaned up that duplication. We just need the variables in the `Coupling` type. The standard name for the Intdiag variables contains `_flipped` which is not true - the copy statements in the stochastic physics wrapper clearly do the same for the Coupling and the Intdiag variables.

Please try to remove lines 6552-6561 and everything else that uses those variables.

JeffBeck-NOAA replied:
Removed all references to variable duplicates using the Intdiag type and confirmed the variables exist in the Coupling type.

climbfuji commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -6971,6 +7037,81 @@
   dimensions = (horizontal_loop_extent)
   type = real
   kind = kind_phys
+[skebu_wts]
```
There is no flipping of variables in the code in `stochastic_physics_wrapper.F90`

JeffBeck-NOAA replied:
I'm not sure where these crept into the code.  I think they're legacy entries from a while ago.  Thanks for finding them!  Removed.

JeffBeck-NOAA replied:
Removed.

climbfuji reviewed (changes requested):
Submodule pointer for upp is out of date - is this branch updated to latest version of fv3atm? Duplication of variables in Intdiag type should be removed. Regarding the regression test b4b differences, please see my comment in the ufs-weather-model PR.

JeffBeck-NOAA committed:
Update submodule pointers

JeffBeck-NOAA committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

JeffBeck-NOAA commented:
> Submodule pointer for upp is out of date - is this branch updated to latest version of fv3atm? Duplication of variables in Intdiag type should be removed. Regarding the regression test b4b differences, please see my comment in the ufs-weather-model PR.

All submodule pointers should now be up-to-date and the PR should now be up-to-date with the latest fv3atm code.

JeffBeck-NOAA committed:
Requested modifications from code review.

JeffBeck-NOAA committed:
Remove "flipped" entries for stochastic physics.

JeffBeck-NOAA committed:
Remove print statements.

JeffBeck-NOAA committed:
Change "v" to "n" to conform with integer variables.

JeffBeck-NOAA committed:
Remove variable duplications in the Intdiag type.

JeffBeck-NOAA committed:
Remove references to Intdiag variables.

JeffBeck-NOAA commented:
@climbfuji, I believe I've addressed all of your concerns except for the question regarding standard names for the newly defined variables.  Hopefully they're OK, but I'll wait for @grantfirl to take a look.

JeffBeck-NOAA commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -2296,6 +2296,71 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
+      allocate (ExtDiag(idx)%data(nblks))
+      do nb = 1,nblks
+        ExtDiag(idx)%data(nb)%var3 => IntDiag(nb)%spp_wts_pbl(:,:)
+      enddo
```
@climbfuji, are these "Intdiag" references OK here? (Full disclosure: I'm trying to shepherd all these code changes that Laurie Carson implemented into a final set of SPP PRs, so I'm learning things myself!)

climbfuji replied:
These `IntDiag` references need to be replaced with `Coupling` (lines 2308, 2321, 2334, 2347, 2360). I'll add suggestions that you can accept.

JeffBeck-NOAA replied:
Thanks!  Fixed!

climbfuji commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -1651,6 +1661,16 @@ module GFS_typedefs
     real (kind=kind_phys), pointer :: tdomip (:)     => null()   !< dominant accumulated sleet type
     real (kind=kind_phys), pointer :: tdoms  (:)     => null()   !< dominant accumulated snow type
 
+    real (kind=kind_phys), pointer :: skebu_wts(:,:) => null()   !< 10 meter u wind speed
```
Lines 1664 to 1673 still need to be removed

JeffBeck-NOAA replied:
Removed!

climbfuji commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -2296,6 +2296,71 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
+      ExtDiag(idx)%mod_name = 'gfs_phys'
+      allocate (ExtDiag(idx)%data(nblks))
+      do nb = 1,nblks
+        ExtDiag(idx)%data(nb)%var3 => IntDiag(nb)%spp_wts_pbl(:,:)
```
```suggestion
        ExtDiag(idx)%data(nb)%var3 => Coupling(nb)%spp_wts_pbl(:,:)
```

climbfuji commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -2296,6 +2296,71 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
+      ExtDiag(idx)%mod_name = 'gfs_phys'
+      allocate (ExtDiag(idx)%data(nblks))
+      do nb = 1,nblks
+        ExtDiag(idx)%data(nb)%var3 => IntDiag(nb)%spp_wts_sfc(:,:)
```
```suggestion
        ExtDiag(idx)%data(nb)%var3 => Coupling(nb)%spp_wts_sfc(:,:)
```

climbfuji commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -2296,6 +2296,71 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
+      ExtDiag(idx)%mod_name = 'gfs_phys'
+      allocate (ExtDiag(idx)%data(nblks))
+      do nb = 1,nblks
+        ExtDiag(idx)%data(nb)%var3 => IntDiag(nb)%spp_wts_mp(:,:)
```
```suggestion
        ExtDiag(idx)%data(nb)%var3 => Coupling(nb)%spp_wts_mp(:,:)
```

climbfuji commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -2296,6 +2296,71 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
+      ExtDiag(idx)%mod_name = 'gfs_phys'
+      allocate (ExtDiag(idx)%data(nblks))
+      do nb = 1,nblks
+        ExtDiag(idx)%data(nb)%var3 => IntDiag(nb)%spp_wts_gwd(:,:)
```
```suggestion
        ExtDiag(idx)%data(nb)%var3 => Coupling(nb)%spp_wts_gwd(:,:)
```

climbfuji commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -2296,6 +2296,71 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
+      ExtDiag(idx)%mod_name = 'gfs_phys'
+      allocate (ExtDiag(idx)%data(nblks))
+      do nb = 1,nblks
+        ExtDiag(idx)%data(nb)%var3 => IntDiag(nb)%spp_wts_rad(:,:)
```
```suggestion
        ExtDiag(idx)%data(nb)%var3 => Coupling(nb)%spp_wts_rad(:,:)
```

climbfuji commented on stochastic_physics/stochastic_physics_wrapper.F90:
```diff
@@ -209,6 +215,32 @@ subroutine stochastic_physics_wrapper (GFS_Control, GFS_Data, Atm_block, ierr)
             end do
          end if
+         if (GFS_Control%do_spp) then
+            DO v=1,GFS_Control%n_var_spp
```
```suggestion
            do n=1,GFS_Control%n_var_spp
```

climbfuji commented on stochastic_physics/stochastic_physics_wrapper.F90:
```diff
@@ -209,6 +215,32 @@ subroutine stochastic_physics_wrapper (GFS_Control, GFS_Data, Atm_block, ierr)
+                     GFS_Data(nb)%Coupling%spp_wts_rad(:,:) = spp_wts(nb,1:GFS_Control%blksz(nb),:,n)
+                 end do
+               end select
+            ENDDO
```
```suggestion
            end do
```

climbfuji reviewed (changes requested):
I noticed the `upp` submodule pointer is still not correct. There should be no changes in the `upp` submodule. And a few other small changes left. We are almost there!

JeffBeck-NOAA committed:
Add the FV3_RRFS_v1alpha SDF back into fv3atm.

JeffBeck-NOAA committed:
Update stochastic_physics/stochastic_physics_wrapper.F90

Co-authored-by: Dom Heinzeller <dom.heinzeller@icloud.com>

JeffBeck-NOAA committed:
Update stochastic_physics/stochastic_physics_wrapper.F90

Co-authored-by: Dom Heinzeller <dom.heinzeller@icloud.com>

JeffBeck-NOAA committed:
Update ccpp/driver/GFS_diagnostics.F90

Co-authored-by: Dom Heinzeller <dom.heinzeller@icloud.com>

JeffBeck-NOAA committed:
Update ccpp/driver/GFS_diagnostics.F90

Co-authored-by: Dom Heinzeller <dom.heinzeller@icloud.com>

JeffBeck-NOAA committed:
Update ccpp/driver/GFS_diagnostics.F90

Co-authored-by: Dom Heinzeller <dom.heinzeller@icloud.com>

JeffBeck-NOAA committed:
Update ccpp/driver/GFS_diagnostics.F90

Co-authored-by: Dom Heinzeller <dom.heinzeller@icloud.com>

JeffBeck-NOAA committed:
Update ccpp/driver/GFS_diagnostics.F90

Co-authored-by: Dom Heinzeller <dom.heinzeller@icloud.com>

JeffBeck-NOAA committed:
Requested changes during code review.

JeffBeck-NOAA committed:
Merge branch 'feature/stoch_spp' of https://github.com/JeffBeck-NOAA/fv3atm into feature/stoch_spp

JeffBeck-NOAA committed:
Update submodule pointers

JeffBeck-NOAA commented:
> I noticed the `upp` submodule pointer is still not correct. There should be no changes in the `upp` submodule. And a few other small changes left. We are almost there!

A "git submodule status" in my feature/stoch_spp branch of fv3atm shows:

 9e1a10f46ec402854b5f921c9ffd092bbb3570c1 upp (upp_gfsv16_release.v1.0.8-110-g9e1a10f)

I think that's the latest, right?

climbfuji commented:
> > I noticed the `upp` submodule pointer is still not correct. There should be no changes in the `upp` submodule. And a few other small changes left. We are almost there!
> 
> A "git submodule status" in my feature/stoch_spp branch of fv3atm shows:
> 
> 9e1a10f46ec402854b5f921c9ffd092bbb3570c1 upp (upp_gfsv16_release.v1.0.8-110-g9e1a10f)
> 
> I think that's the latest, right?

I don't think it's pointing to the head of the branch. fv3atm currently is pointing to hash c939eae of UPP, see https://github.com/NOAA-EMC/fv3atm/tree/96a2554bf20bb340f24288406f1d0d02eb9bbb5b.

JeffBeck-NOAA commented:
> > > I noticed the `upp` submodule pointer is still not correct. There should be no changes in the `upp` submodule. And a few other small changes left. We are almost there!
> > 
> > 
> > A "git submodule status" in my feature/stoch_spp branch of fv3atm shows:
> > 9e1a10f46ec402854b5f921c9ffd092bbb3570c1 upp (upp_gfsv16_release.v1.0.8-110-g9e1a10f)
> > I think that's the latest, right?
> 
> I don't think it's pointing to the head of the branch. fv3atm currently is pointing to hash c939eae of UPP, see https://github.com/NOAA-EMC/fv3atm/tree/96a2554bf20bb340f24288406f1d0d02eb9bbb5b.

c939eae is a commit from November 10th, so it looks like fv3atm is not up-to-date.  See here: https://github.com/NOAA-EMC/UPP/commits/develop

Edit:  Sorry, misunderstood!  I will need to find out how to set a pointer to a specific hash.  I'll update soon.

Edit2:  Should be good now!

JeffBeck-NOAA committed:
Update UPP submodule pointer

JeffBeck-NOAA committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

JeffBeck-NOAA committed:
Update ccpp-physics submodule pointer

JeffBeck-NOAA committed:
Set upp submodule pointer to specific hash in fv3atm.

JeffBeck-NOAA committed:
Set UPP submodule pointer to specific hash in fv3atm.

JeffBeck-NOAA committed:
Merge branch 'feature/stoch_spp' of https://github.com/JeffBeck-NOAA/fv3atm into feature/stoch_spp

junwang-noaa commented:
@JeffBeck-NOAA There is a UPP PR #449 to update the UPP, which will update the UPP version, you may need to update your branch when that PR is committed.

JeffBeck-NOAA commented:
> @JeffBeck-NOAA There is a UPP PR #449 to update the UPP, which will update the UPP version, you may need to update your branch when that PR is committed.

@junwang-noaa, thank you for the heads up!  I will be on the lookout for that PR to be merged.

JeffBeck-NOAA committed:
Update submodule pointer for ccpp/physics

climbfuji commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -1161,6 +1166,11 @@ module GFS_typedefs
                                               ! we'd need to make this dim(6,5).
+    logical              :: do_spp
+    integer              :: n_var_spp
+    character(len=3)     :: spp_var_list(6)  ! dimension here must match  n_var_max_spp in  stochy_nml_def
```
I would recommend to NOT hardcode the length of the `spp_var_list` array here, but follow the logic for `lndp_var_list`. As far as I can see, both `spp_var_list` and `lndp_var_list` are handled in the same way: allocated in GFS_typedefs, but the namelist read happens in the stochastic_physics code. Same for spp_prt_list.

JeffBeck-NOAA replied:
Fixed.

JeffBeck-NOAA committed:
Changes for PR modification requests

JeffBeck-NOAA committed:
Changes for PR modifcation requests

JeffBeck-NOAA committed:
Test changes after PR review modifications

JeffBeck-NOAA committed:
Test changes after PR review modifications

JeffBeck-NOAA committed:
Allocate spp_wts_* arrays in GFS_typedefs.F90 and not in the stochastic physics wrapper.

JeffBeck-NOAA committed:
Allocate spp_wts_* arrays in GFS_typedefs.F90 and not in the stochastic physics wrapper.

JeffBeck-NOAA committed:
Merge branch 'feature/stoch_spp_part2' of https://github.com/JeffBeck-NOAA/fv3atm into feature/stoch_spp_part2

JeffBeck-NOAA committed:
Update pointers for ccpp/physics and ccpp/framework

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Change pointers and .gitmodule feature branch for ccpp-physics

JeffBeck-NOAA committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

JeffBeck-NOAA committed:
Update pointer for ccpp/physics

JeffBeck-NOAA committed:
Update UPP pointer

JeffBeck-NOAA committed:
Update UPP pointer

JeffBeck-NOAA committed:
Don't limit the size of spp_var_list

JeffBeck-NOAA committed:
Update pointer for ccpp/physics

JeffBeck-NOAA committed:
Update pointer for ccpp/physics

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Fix spp_var/prt_list in GFS_typedefs.F90

JeffBeck-NOAA committed:
Make SPP variables pointers in GFS_typedefs.F90

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -2244,6 +2244,46 @@
   kind = kind_phys
   active = (flag_for_stochastic_skeb_option)
+[spp_wts_pbl]
+  standard_name = weights_for_stochastic_spp_pbl_perturbations
```
standard_name suggestion: spp_weights_for_pbl_scheme

JeffBeck-NOAA replied:
Updated.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -2244,6 +2244,46 @@
+  kind = kind_phys
+  active = (flag_for_stochastic_spp_option)
+[spp_wts_sfc]
+  standard_name = weights_for_stochastic_spp_sfc_perturbations
```
standard_name suggestion: spp_weights_for_surface_layer_scheme

JeffBeck-NOAA replied:
Updated.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -2244,6 +2244,46 @@
+  kind = kind_phys
+  active = (flag_for_stochastic_spp_option)
+[spp_wts_mp]
+  standard_name = weights_for_stochastic_spp_mp_perturbations
```
standard_name suggestion: spp_weights_for_microphysics_scheme

JeffBeck-NOAA replied:
Updated.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -2244,6 +2244,46 @@
+  kind = kind_phys
+  active = (flag_for_stochastic_spp_option)
+[spp_wts_gwd]
+  standard_name = weights_for_stochastic_spp_gwd_perturbations
```
standard_name suggestion: spp_weights_for_gravity_wave_drag_scheme

JeffBeck-NOAA replied:
Updated.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -2244,6 +2244,46 @@
+  kind = kind_phys
+  active = (flag_for_stochastic_spp_option)
+[spp_wts_rad]
+  standard_name = weights_for_stochastic_spp_rad_perturbations
```
standard_name suggestion: spp_weights_for_radiation_scheme

JeffBeck-NOAA replied:
Updated.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -4610,6 +4650,12 @@
   dimensions = ()
   type = logical
+[do_spp]
+  standard_name = flag_for_stochastic_spp_option
```
standard_name suggestion: do_stochastically_perturbed_parameterizations

The CCPP community has settled on "do_..." rather than flag. This hasn't propagated to the UFS yet, but we can start using it. See, e.g. https://github.com/ESCOMP/CCPPStandardNames/blob/main/standard_names.xml

JeffBeck-NOAA replied:
Updated.  Thanks!

junwang-noaa replied:
@JeffBeck-NOAA it seems to me the name is not updated to "do_stochastically_perturbed_parameterizations"?

JeffBeck-NOAA replied:
Updated.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -4638,6 +4684,56 @@
   kind = len=3
   active = (control_for_stochastic_land_surface_perturbation /= 0)
+[n_var_spp]
+  standard_name = number_of_spp_schemes_perturbed
```
standard_name suggestion: number_of_perturbed_spp_schemes

JeffBeck-NOAA replied:
Updated.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -4638,6 +4684,56 @@
+  dimensions = ()
+  type = integer
+[spp_prt_list]
+  standard_name =magnitude_of_perturbations_for_spp
```
standard_name suggestion: magnitude_of_spp_perturbations

JeffBeck-NOAA replied:
Updated.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -4638,6 +4684,56 @@
+  type = real
+  kind = kind_phys
+[spp_var_list]
+  standard_name = variables_to_be_perturbed_for_spp
```
Why is this called "variables"? It seems like it is referring to a scheme type (pbl, rad, etc.). If so:

standard_name suggestion: perturbed_spp_schemes

JeffBeck-NOAA replied:
Updated.  This should indeed be schemes, not variables.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -2244,6 +2244,46 @@
+[spp_wts_pbl]
+  standard_name = weights_for_stochastic_spp_pbl_perturbations
+  long_name = weights for stochastic spp pbl perturbations
+  units = none
```
For this and other spp_wts_*, units = 1 should replace units = none, which, according to the rules needed by the parser. See the bottom section talking about units for dimensionless variables: https://github.com/ESCOMP/CCPPStandardNames/blob/main/StandardNamesRules.rst

JeffBeck-NOAA replied:
Updated.  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -4638,6 +4684,56 @@
+[spp_prt_list]
+  standard_name =magnitude_of_perturbations_for_spp
+  long_name = magnitude of perturbations for spp
+  units = variable
```
I think that this needs to be "units = 1". "Units = variable" may cause trouble in the future if not now. I realize that this was probably following the land surface perturbations, but we'll probably need to clean metadata associated with that up in the future too.

JeffBeck-NOAA replied:
Updated.  Thanks!

junwang-noaa replied:
Same here. The unit is still "variable"

JeffBeck-NOAA replied:
Fixed!  Thanks!

grantfirl commented on ccpp/data/GFS_typedefs.meta:
```diff
@@ -4638,6 +4684,56 @@
+  type = character
+  kind = len=3
+[spp_pbl]
+  standard_name = control_for_planetary_boundary_layer_spp_perturbations
```
It looks like most of the names have been using "pbl" instead of "planetary_boundary_layer". It doesn't exist on our list of abbreviations, but it should. PBL is well known in the community, so this'll save some characters.

JeffBeck-NOAA replied:
Updated.  Thanks!

JeffBeck-NOAA committed:
Updates to standard names and units

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Change spp_prt_list to a variable unit

JeffBeck-NOAA committed:
Update do_spp metadata

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

JeffBeck-NOAA committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Update units for spp_prt_list.

JeffBeck-NOAA committed:
Update do_spp standard name.

JeffBeck-NOAA committed:
Update do_spp standard name

JeffBeck-NOAA committed:
Revert erroneous do_sppt standard name change.

JeffBeck-NOAA committed:
Change all references of flag_for_stochastic_spp_option to do_stochastically_perturbed_parameterizations

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Update ccpp/physics pointer

JeffBeck-NOAA committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into feature/stoch_spp

JeffBeck-NOAA committed:
Update ccpp/physics pointer

junwang-noaa commented:
@climbfuji Would you please review the PR again? We can't commit the PR with "change requested" review comments.

climbfuji commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -2296,6 +2296,71 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
+      enddo
+    endif
+    
+    if (Model%do_spp) then
```
Future cleanup. All these individual `if (Model%do_spp) then` blocks could be combined into one.

JeffBeck-NOAA committed:
Update ccpp-physics pointer and .gitmodules file

binli2337 commented:
@junwang-noaa I am not able to merge this pull request.

binli2337 commented:
@JeffBeck-NOAA The ccpp-physics branch needs to be changed to main.

junwang-noaa commented on .gitmodules:
```diff
@@ -9,7 +9,7 @@
 [submodule "ccpp/physics"]
   path = ccpp/physics
   url = https://github.com/NCAR/ccpp-physics
-  branch = main
```
@JeffBeck-NOAA Please change the branch back to the main branch. Thanks

JeffBeck-NOAA replied:
Fixed! Thank you!

JeffBeck-NOAA committed:
Change branch back to main
